ert_shared/models/base_run_model.py

```diff
@@ -319,6 +326,136 @@ def run_ensemble_evaluator(
+
+            await ensemble.evaluate_async(ee_config, self.id_)
+
+            await ensemble_listener
+
+            for iens, run_arg in enumerate(run_context):
+                if run_context.is_active(iens):
+                    if run_arg.run_status in (
+                        RunStatusType.JOB_LOAD_FAILURE,
+                        RunStatusType.JOB_RUN_FAILURE,
+                    ):
+                        run_context.deactivate_realization(iens)
+
+            await loop.run_in_executor(
+                pool,
+                run_context.get_sim_fs().fsync,
+            )
+
+    @abstractmethod
+    async def run(self, evaluator_server_config: EvaluatorServerConfig) -> None:
+        raise NotImplementedError
+
+    async def successful_realizations(self, iter_: int) -> int:
+        return self._state_machine.successful_realizations(iter_)
+
+    async def _run_hook(
+        self,
+        hook: int,  # HookRuntime
+        iter_: int,
+        loop: asyncio.AbstractEventLoop,
+        executor: concurrent.futures.Executor,
+    ) -> None:
+        # Send HOOK_STARTED
+        await self.dispatch(
+            CloudEvent(
+                {
+                    "type": identifiers.EVTYPE_EXPERIMENT_HOOK_STARTED,
+                    "source": f"/ert/experiment/{self.id_}",
+                    "id": str(uuid.uuid1()),
+                },
+                {
+                    "name": str(hook),
+                },
+            ),
+            iter_,
+        )
+
+        # Run hook
+        await loop.run_in_executor(
+            executor,
+            EnkfSimulationRunner.runWorkflows,
+            hook,
+            self.ert(),
+        )
+
+        # Send HOOK_ENDED
+        await self.dispatch(
+            CloudEvent(
+                {
+                    "type": identifiers.EVTYPE_EXPERIMENT_HOOK_ENDED,
+                    "source": f"/ert/experiment/{self.id_}",
+                    "id": str(uuid.uuid1()),
+                },
+                {
+                    "name": str(hook),
+                },
+            ),
+            iter_,
+        )
+
+    @property
+    def id_(self) -> str:
+        if not self._id:
+            raise RuntimeError(f"{self} does not have an ID")
+        return self._id
+
+    @id_.setter
+    def id_(self, value: str) -> None:
+        if self._id is not None:
+            raise ValueError("experiment id can only be set once")
+        self._id = value
+
+    async def _ensemble_listener(self, ensemble: Ensemble, iter_: int) -> None:
+        """Redirect events emitted by the ensemble to this experiment."""
+        while True:
+            event: CloudEvent = await ensemble.output_bus.get()
+            await self.dispatch(event, iter_)
+            if event["type"] in (
+                identifiers.EVTYPE_ENSEMBLE_FAILED,
+                identifiers.EVTYPE_ENSEMBLE_CANCELLED,
+                identifiers.EVTYPE_ENSEMBLE_STOPPED,
+            ):
+                break
```

Comment:
Right, so here we don't care the reason it stopped but just that it stops - just clearing my thoughts :)

Reply:
Correct!